### PR TITLE
changes to `repeat` so it can be more general

### PR DESCRIPTION
**Pull request by EderSantana**
Add a new parameter `dim` with default 1 to the `repeat` function.

### CODE REVIEW

**Comment by EderSantana** [comment]
A few things: 
* `repeat` was not general enough.
* `expand_dims` uses `dim` to what you meant `axis` everywhere else in the backend.

**Comment by fchollet** [comment]
- please update the docstrings as well
- the changes do not work as expected, try running `tests/test_shape_inference.py`.

**Comment by fchollet** [comment]
Maybe check out what `layers.convolutional.UpSampling1D/2D` does.

**Comment by EderSantana** [comment]
Do you agree with the renaming dim -> axis?

**Comment by fchollet** [comment]
> Do you agree with the renaming dim -> axis?

Sure.

> Btw, travis doesn't work? Is there a way to install TF on it?

It's on my to-do list. This will be up before the release, maybe even today.

**Comment by EderSantana** [comment]
Things like `W[:, None]` are not valid in TF? We alway have to use `expand_dims`? If so, we have to think of a way to make `expand_dim` to accept a list of inputs otherwise it may be too repetitive.

**Comment by fchollet** [comment]
Repetitiveness and verbosity are largely fine. What matters most is robustness, simplicity and clarity.

**Comment by EderSantana** [comment]
ok, I'll go back to converting NTM, I just wanted to be sure of these things

**Comment by EderSantana** [comment]
I have another problem now, `repeat` automatically creates a new dimension, this is not compatible with Theano.tensor.repeat for example. I need a way to repeat a new tensor in a dimension that already exists, what should I do? Create a new function in the backends, change the repeat, add an option into repeat? 

I'm already running tests in the NTM translation. Keras's new step+rnn is so clean and beautiful it makes me happy.

**Comment by fchollet** [comment]
> I have another problem now, repeat automatically creates a new dimension, this is not compatible with Theano.tensor.repeat for example. 

`K.repeat` performs a very specific function that is different from `T.repeat`. Think of `K.repeat` as the backend op for the `RepeatVector` layer. 

Most of the time you probably don't want to use `K.repeat`. Maybe we should rename it to `repeat_vector` for added clarity.

> I need a way to repeat a new tensor in a dimension that already exists, what should I do? Create a new function in the backends, change the repeat, add an option into repeat?

I don't think there is any need for a new operation. Simply `K.concatenate` a tensor with itself along a specific axis, much like what happens in `UpSampling1D`.

**Comment by fchollet** [comment]
Will you add the NTM changes in this PR?

**Comment by EderSantana** [comment]
I got you. Well for my problem, I was nesting expand_dims with repeat. I just have to use repeat alone. 

One thing: Do your think something like `K.concatenate([x]*K.shape(a_tensor)[0], axis=0)` will work? That shape value may be symbolic. But still, If we face that problem in the future, we think about a solution, back to what we have now.

File: tests/auto/keras/layers/test_recurrent.py [hunk 1]
```diff
@@ -42,6 +42,9 @@ def test_gru(self):
     def test_lstm(self):
         _runner(recurrent.LSTM)
 
+    def test_ntm(self):
```
Comment (EderSantana) [inline tests/auto/keras/layers/test_recurrent.py]: @fchollet you can comment out this test and accept everything else, I may need some more time to work on the NTM. Shouldn't be long.

**Comment by fchollet** [comment]
It won't work. I believe it would work it Theano but it wouldn't in TensorFlow. 

Because shape handling is very different from Theano to TF, shapes can generally never be used in the code. In this case the samples dimension (shape[0]) simply cannot be accessed in TF code, because it is dynamic and TF shape handling is *not* symbolic, but instead explicit. You would just get 'None' if you tried to access it. So you have to build around it.

**Comment by fchollet** [comment]
By the way, do you think the NTM code should live in `examples/` entirely, or do you see sufficient reusability to put it in `keras/layers/ntm.py`?

**Comment by EderSantana** [comment]
we can use the NTM as a regular RNN layer. For ex. there is a recent paper on ICLR about stacking NTMs. But I'll be fine either way. 

Just one update on the code, I'm having little difficulties with what is happing inside the recurrent part, mainly getting used to the new way of defining initial states. I'll debug it later this week, because tomorrow is my thesis proposal and I got a little busy this weekend.

But as I said, you can remove the `ntm_test` I introduced in this PR and accept the other things.

**Comment by EderSantana** [comment]
So I messed this PR up. I'll start a new one.
